core/src/main/java/nl/nn/adapterframework/jta/narayana/NarayanaDataSourceFactory.java

```diff
@@ -16,61 +16,96 @@
 
 	@Override
 	protected DataSource augmentDatasource(CommonDataSource dataSource, String dataSourceName) {
-		if (dataSource instanceof XADataSource) {
+		if(dataSource instanceof XADataSource) {
 			XAResourceRecoveryHelper recoveryHelper = new DataSourceXAResourceRecoveryHelper((XADataSource) dataSource);
 			this.transactionManager.registerXAResourceRecoveryHelper(recoveryHelper);
-			NarayanaDataSource result = new NarayanaDataSource(dataSource, dataSourceName);
-			result.setMaxConnections(maxPoolSize);
-			checkModifiers(result);
-			return result;
+
+			if(maxPoolSize > 1) {
+				return XAPool((XADataSource) dataSource);
+			}
+
+			NarayanaDataSource ds = new NarayanaDataSource(dataSource, dataSourceName);
+			log.info("registered Narayana DataSource [{}] with Transaction Manager", ds);
+			return ds;
 		}
 
-		log.warn("DataSource [{}] is not XA enabled", dataSourceName);
-		return (DataSource) dataSource;
+		log.info("DataSource [{}] is not XA enabled, unable to register with an Transaction Manager", dataSourceName);
+		return pool((DataSource) dataSource);
 	}
 
-	public static void checkModifiers(DataSource dataSource) {
-		try (Connection connection = dataSource.getConnection()) {
-			DatabaseMetaData metadata = connection.getMetaData();
-			String driverName = metadata.getDriverName();
-			int major = metadata.getDriverMajorVersion();
-			int minor = metadata.getDriverMinorVersion();
+	private DataSource pool(DataSource dataSource) {
+		ConnectionFactory cf = new DataSourceConnectionFactory(dataSource);
+		PoolableConnectionFactory poolableConnectionFactory = new PoolableConnectionFactory(cf, null);
 
-			if (ModifierFactory.getModifier(driverName, major, minor)==null) {
-				LOG.info("No Modifier found for driver [{}] version [{}.{}], creating IsSameRM modifier", driverName, major, minor);
-				ModifierFactory.putModifier(driverName, major, minor, IsSameRMModifier.class.getName());
-			}
-		} catch (SQLException e) {
-			LOG.warn("Could not check for existence of Modifier", e);
-		}
+		ObjectPool<PoolableConnection> connectionPool = createConnectionPool(poolableConnectionFactory);
+
+		PoolingDataSource<PoolableConnection> ds = new PoolingDataSource<>(connectionPool);
+		log.info("created PoolingDataSource [{}]", ds);
+		return ds;
+	}
+
+	private DataSource XAPool(XADataSource dataSource) {
+		XAConnectionFactory cf = new DataSourceXAConnectionFactory(transactionManager.getTransactionManager(), dataSource);
+		PoolableConnectionFactory poolableConnectionFactory = new PoolableManagedConnectionFactory(cf, null);
+
+		ObjectPool<PoolableConnection> connectionPool = createConnectionPool(poolableConnectionFactory);
+
+		PoolingDataSource<PoolableConnection> ds = new ManagedDataSource<>(connectionPool, cf.getTransactionRegistry());
+		log.info("created XA-enabled PoolingDataSource [{}]", ds);
+		return ds;
 	}
 
+	private ObjectPool<PoolableConnection> createConnectionPool(PoolableConnectionFactory poolableConnectionFactory) {
```

Comment:
This method is duplicated with the code to create pool for BTM -- can this code perhaps be shared? To ease maintenance if something should still be changed ?

Or do you think that soon enough we'll be able to drop BTM and then better to have all the relevant code here in this class?

Reply:
No, that's all BTM native, and thus it cannot be shared.

Reply:
Yea I know it's confusing, but now we know where BTM has it's pooling mechanism from 😄